forc-plugins/forc-fmt/src/main.rs

```diff
@@ -63,35 +76,178 @@ fn run() -> Result<()> {
+        if let Ok(read_dir) = fs::read_dir(next_dir) {
+            for entry in read_dir.filter_map(|res| res.ok()) {
+                let path = entry.path();
+
+                if path.is_dir() {
+                    dirs_to_search.push(path.clone());
+                    if path.join(constants::MANIFEST_FILE_NAME).exists() {
+                        dirs_to_format.push(path);
+                    }
+                }
+            }
+        }
+    }
+
+    dirs_to_format
+}
+
+/// Format a file, given its path.
+/// Returns:
+/// - Ok(true) if executed successfully and formatted,
+/// - Ok(false) if executed successfully and not formatted,
+/// - Err if it fails to execute at all.
+fn format_file(
+    app: &App,
+    file: PathBuf,
+    manifest_file: Option<PathBuf>,
+    formatter: &mut Formatter,
+) -> Result<bool> {
+    let file = file.canonicalize()?;
+    if let Ok(file_content) = fs::read_to_string(&file) {
+        let mut edited = false;
+        let file_content: Arc<str> = Arc::from(file_content);
+        let build_config = manifest_file.map(|f| {
+            BuildConfig::root_from_file_name_and_manifest_path(
+                file.clone(),
+                f,
+                BuildTarget::default(),
+            )
+        });
+        match Formatter::format(formatter, file_content.clone(), build_config.as_ref()) {
             Ok(formatted_content) => {
-                format_file(path, &formatted_content)?;
+                if app.check {
+                    if *file_content != formatted_content {
+                        info!("\n{:?}\n", file);
+                        display_file_diff(&file_content, &formatted_content)?;
+                        edited = true;
+                    }
+                } else {
+                    write_file_formatted(&file, &formatted_content)?;
+                }
+
+                return Ok(edited);
             }
             Err(err) => {
+                // there could still be Sway files that are not part of the build
+                error!("\nThis file: {:?} is not part of the build", file);
                 error!("{}\n", err);
             }
         }
     }
 
-    Ok(())
+    bail!("Could not read file")
+}
+
+/// Format the workspace at the given directory.
+fn format_workspace_at_dir(app: &App, workspace: &WorkspaceManifestFile, dir: &Path) -> Result<()> {
+    let mut contains_edits = false;
+    let mut formatter = Formatter::from_dir(dir)?;
+    let mut members = vec![];
+
+    for member_path in workspace.member_paths()? {
+        members.push(member_path)
+    }
+
+    // Format files at the root - we do not want to start calling format_pkg_at_dir() here,
+    // since this would mean we format twice on each subdirectory.
+    if let Ok(read_dir) = fs::read_dir(dir) {
+        for entry in read_dir.filter_map(|res| res.ok()) {
+            let path = entry.path();
+            if path.is_file() && is_sway_file(&path) {
```

Comment:
Consider moving the `is_file` check inside `is_sway_file`, since it seems like we'd always want to check these together:

```
pub fn is_sway_file(file: &Path) -> bool {
    file.is_file() && Some(OsStr::new(constants::SWAY_EXTENSION)) == file.extension()
}
```